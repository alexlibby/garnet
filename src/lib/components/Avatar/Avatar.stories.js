import Avatar from "../Avatar/Avatar.svelte";
import { BADGE } from "storybook-addon-badges";

const BADGES = {
  ...BADGE,
  ALPHA: "Alpha",
};

export default {
  title: "Garnet UI Library/React Components/Avatar",
  component: Avatar,
  props: {
    status: "available",
    src: "/avatar.png",
  },
};

export const Default = () => ({
  Component: Avatar,
  props: {
    statusSize: "small",
    src: "/avatar.png",
  },
});

Default.parameters = {
  badges: [BADGES.ALPHA],
};

export const Busy = () => ({
  Component: Avatar,
  props: {
    status: "busy",
    src: "/avatar.png",
  },
});

Busy.parameters = {
  badges: [BADGES.ALPHA],
};

export const Away = () => ({
  Component: Avatar,
  props: {
    status: "away",
    statusSize: "large",
    src: "/avatar.png",
  },
});

Away.parameters = {
  badges: [BADGES.ALPHA],
};

export const Unavailable = () => ({
  Component: Avatar,
  props: {
    status: "unavailable",
    statusSize: "medium",
    src: "/avatar.png",
  },
});

Unavailable.parameters = {
  badges: [BADGES.ALPHA],
};
