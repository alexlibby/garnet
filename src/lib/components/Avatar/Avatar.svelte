<svelte:options customElement="garnet-avatar" />

<script>
  export let src = "";
  export let status = "available" || "busy" || "away" || "unavailable";
  export let statusSize = "small" || "medium" || "large";

  $: statusClasses = () => {
    let inputClasses = [status, statusSize];
    inputClasses = inputClasses.filter((klass) => klass.length);
    return inputClasses.join(" ");
  };
</script>

<div class="garnet-avatar">
  {#if src !== null}
    <img {src} class="avatar-img" alt="avatar" />
  {/if}
  <slot />
  <span data-testid="availability" class={["base", statusClasses()].join(" ")} />
</div>

<style>
.garnet-avatar {
    position: relative;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    vertical-align: middle;
    border-radius: 800px;
    width: 32px;
    height: 32px;
    font-size: 12.8px;
  }

  .base {
    border-radius: 800px;
    position: absolute;
    border: 1px solid #373737;
  }  

  .small {
    width: 8px;
    height: 8px;
    top: 25px;
    right: 0px;
  }

  .medium {
    width: 12px;
    height: 12px;
    top: 22px;
    right: -4px;
  }

  .large {
    width: 16px;
    height: 16px;
    top: 22px;
    right: -4px;
  }

  .available { background-color: #00ff00; }
  .busy { background-color: #ff0000; }
  .away { background-color: #ffff00; }
  .unavailable { background-color: #ffffff; border: 1px solid #000000;}
  .avatar-img { width: 100%; height: auto; border-radius: 800px; }
</style>
  