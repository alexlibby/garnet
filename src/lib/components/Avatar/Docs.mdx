import { Canvas, Meta } from "@storybook/addon-docs";
import { ArgTypes } from "@storybook/blocks";
import * as AvatarStories from "./Avatar.stories";

<Meta of={AvatarStories} />

# Avatar

Avatar is the primary component. It has four possible states.

- [Default](#default)
- [Busy](#busy)
- [Away](#away)
- [Unavailable](#unavailable)

<a id="default" />

## Default

This is the initial version of the Avatar component.

<Canvas of={AvatarStories.Default} />

<a id="busy" />

## Busy

This variant displays the Avatar component styled with a red status.

<Canvas of={AvatarStories.Busy} />

<a id="away" />

## Away

This variant displays the Avatar component styled with a yellow status.

<Canvas of={AvatarStories.Away} />

<a id="unavailable" />

## Unavailable

This variant displays the Avatar component with a white status indicator.

<Canvas of={AvatarStories.Unavailable} />

## Properties of component

Below is a list of arguments available for this component:

<ArgTypes of={AvatarStories} />
