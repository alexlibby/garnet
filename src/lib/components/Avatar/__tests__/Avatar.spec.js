import { describe, test, expect } from "vitest";
import { axe } from "vitest-axe";
import * as AxeMatchers from "vitest-axe/matchers";
import { render, screen, fireEvent } from "@testing-library/svelte";
import Avatar from "../../Avatar/Avatar.svelte";

const AvatarHTML = `<div class="garnet-avatar s-q7VgqBOCP_sX"><img src="/avatar.png" class="avatar-img s-q7VgqBOCP_sX" alt="avatar">  <span class="base available small s-q7VgqBOCP_sX"></span></div>`;
expect.extend(AxeMatchers);

describe("Tests for Avatar", () => {
  const mockText = "This is a Avatar";

  const props = {
    status: "available",
    src: "/avatar.png",
    statusSize: "smal",
  };

  test("should render properly", () => {
    const result = render(Avatar, { props });
    expect(() => result).not.toThrow();
  });

  test("show show a label with a red busy status", () => {
    const result = render(Avatar, { ...props, status: "busy" });

    const element = document.querySelector("span");
    // element has a class
    expect(element.classList).toContain("busy");
  });

  test("get a snapshot of component", () => {
    // Create a new container for the test and mount component
    const host = document.createElement("div");
    document.body.appendChild(host);
    const instance = new Avatar({ target: host, props: props });

    // check that all rendered OK
    expect(instance).toBeTruthy();

    // Take snapshot
    expect(host.innerHTML).toMatchSnapshot();
  });

  test("should demonstrate no issues with accessibility", async () => {
    const render = () => AvatarHTML;
    const result = render();

    const results = await axe(result, {
      rules: {
        region: { enabled: false },
      },
    });

    expect(results).toHaveNoViolations();
  });
});
