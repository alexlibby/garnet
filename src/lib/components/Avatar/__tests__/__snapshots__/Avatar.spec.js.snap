// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for Avatar > get a snapshot of component 1`] = `"<div class="garnet-avatar svelte-1bg7u5z"><img src="/avatar.png" class="avatar-img svelte-1bg7u5z" alt="avatar">  <span data-testid="availability" class="base available smal svelte-1bg7u5z"></span></div>"`;
